YatimaStdLib/SparseMatrix.lean

```diff
@@ -0,0 +1,190 @@
+import YatimaStdLib.Array
+import YatimaStdLib.Ord
+import YatimaStdLib.Ring
+import YatimaStdLib.RBMap
+import Std.Data.RBMap
+
+structure SparseMatrix (R : Type _) where
+  entries : Std.RBMap (Nat × Nat) R compare
+  rows : Nat
+  cols : Nat
+
+open Std.RBMap
+
+/--
+Empty matrix
+-/
+def SparseMatrix.empty (col : Nat) (row : Nat) : SparseMatrix R :=
+  SparseMatrix.mk Std.RBMap.empty col row
+
+/--
+Insert a triple
+-/
+def SparseMatrix.insert
+  (col : Nat) (row : Nat) (a : R) (m : SparseMatrix R) : SparseMatrix R :=
+  SparseMatrix.mk
+    (Std.RBMap.insert m.entries (col,row) a)
+    m.rows
+    m.cols
+
+/--
+Creates a sparse matrix from a list
+-/
+def ofList (l : List ((Nat × Nat) × R)) : SparseMatrix R :=
+  let entries := Std.RBMap.ofList l compare
+  SparseMatrix.mk
+    entries
+    (1 + (Array.foldr max 0 $ Array.map (fun p => p.1) $ keysArray entries))
+    (1 + (Array.foldr max 0 $ Array.map (fun p => p.2) $ keysArray entries))
```

Comment:
These should be parameters of the function. Suppose we have a 1000x1000 matrix whose only non-zero value (10) is at (0, 0). That would be `.ofList ((0, 0), 10) 1000 1000`.

Maybe `ofList` is not a good name. What about `ofValuesWithDims` or something similar?

Reply:
Makes sense